base_layer/core/src/proof_of_work/monero_rx.rs

```diff
@@ -231,25 +257,538 @@ fn verify_root(monero_data: &MoneroData) -> Result<(), MergeMineError> {
+            coinbase_tx: block.miner_tx,
+        };
+        let serialized = bincode::serialize(&monero_data).unwrap();
+        let pow = ProofOfWork {
+            accumulated_monero_difficulty: Default::default(),
+            accumulated_blake_difficulty: Default::default(),
+            target_difficulty: Default::default(),
+            pow_algo: PowAlgorithm::Monero,
+            pow_data: serialized,
+        };
+        block_header.pow = pow;
+        let monero_data = MoneroData::new(&block_header).unwrap();
+        verify_header(&block_header, &monero_data).unwrap();
+    }
+
+    #[test]
+    #[should_panic]
+    fn test_append_mm_tag_no_tag() {
+        let blocktemplate_blob = "0c0c8cd6a0fa057fe21d764e7abf004e975396a2160773b93712bf6118c3b4959ddd8ee0f76aad0000000002e1ea2701ffa5ea2701d5a299e2abb002028eb3066ced1b2cc82ea046f3716a48e9ae37144057d5fb48a97f941225a1957b2b0106225b7ec0a6544d8da39abe68d8bd82619b4a7c5bdae89c3783b256a8fa47820208f63aa86d2e857f070000".to_string();
+        let seed_hash = "9f02e032f9b15d2aded991e0f68cc3c3427270b568b782e55fbd269ead0bad97".to_string();
+        let bytes = hex::decode(blocktemplate_blob).unwrap();
+        let mut block = deserialize::<MoneroBlock>(&bytes[..]).unwrap();
+        let mut block_header = BlockHeader {
+            version: 0,
+            height: 0,
+            prev_hash: vec![0],
+            timestamp: Default::default(),
+            output_mr: vec![0],
+            range_proof_mr: vec![0],
+            kernel_mr: vec![0],
+            total_kernel_offset: RistrettoSecretKey::from(0),
+            nonce: 0,
+            pow: ProofOfWork::default(),
+        };
+        let count = 1 + (block.tx_hashes.len() as u16);
+        let mut hashes = Vec::with_capacity(count as usize);
+        let mut proof = Vec::with_capacity(count as usize);
+        hashes.push(block.miner_tx.hash());
+        proof.push(block.miner_tx.hash());
+        for item in block.clone().tx_hashes {
+            hashes.push(item);
+            proof.push(item);
+        }
+        let root = tree_hash(hashes.clone()).unwrap();
+        let monero_data = MoneroData {
+            header: block.header,
+            key: seed_hash,
+            count,
+            transaction_root: from_slice(&root),
+            transaction_hashes: from_hashes(&hashes),
+            coinbase_tx: block.miner_tx,
+        };
+        let serialized = bincode::serialize(&monero_data).unwrap();
+        let pow = ProofOfWork {
+            accumulated_monero_difficulty: Default::default(),
+            accumulated_blake_difficulty: Default::default(),
+            target_difficulty: Default::default(),
+            pow_algo: PowAlgorithm::Monero,
+            pow_data: serialized,
+        };
+        block_header.pow = pow;
+        let monero_data = MoneroData::new(&block_header).unwrap();
+        verify_header(&block_header, &monero_data).unwrap();
+    }
+
+    #[test]
+    #[should_panic]
```

Comment:
Generally speaking, the `#[should_panic]` attribute should almost never be needed. Rather than allowing that the test to pass when panicking (which can happen in all sorts of cases where yo'd want the test to fail) make assertions which will panic if they are not met.

The `unpack_enum!` utility macro in `tari_test_utils` helps a lot with error enums.

```rust
use tari_test_utils::unpack_enum;

// (....)

let err =  verify_header(&block_header, &monero_data).unwrap_err();
// I may prefer adding a new error variant to describe this specific error e.g MergeMineError::MoneroCoinbaseTagNotFound - but up to you
unpack_enum!(MergeMineError::ValidationError(details) = err);

 assert!(details.contains("merge mining tag was not found"));


```